Ледянкин Павел

PavelLediankin opened the pull request.

PavelLediankin committed:
Task1

PavelLediankin commented:
@dmnovikova

PavelLediankin committed:
Task 2

PavelLediankin committed:
Task 3 update T2

В Т2 сделал вывод раскладок в рабочую директорию

PavelLediankin committed:
optimisation CCL + update testes

Уменьшен коэффициент увеличивавший дистанцию при увеличении угла
Скорректировны тесты (по времени и по плотности раскладки)
Добавлена визуализация раскладки с большим разбросом значений

PavelLediankin committed:
Add 4.jpeg to Readme

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+{
+    public class CircularCloudLayouter
+    {
+        private List<RectangleF> rectangles;
```
Зачем такая сложная конструкция?
```private List<RectangleF> rectangles;```
```public List<RectangleF> Rectangles { get => rectangles; }```

Ведь это тоже самое, что ```public List<RectangleF> Rectangles { get; }```

PavelLediankin replied:
Да переделаю, просто в начале у меня где-то здесь была проблема, может опечатался

dmnovikova replied:
тут проблема в том, что никто не мешает извне сделать layouter.Rectangles.Add(...)
но твоя конструкция ее не решает.
Вот если бы ты написал 
```
private List<RectangleF> rectangles;
public IReadOnlyList<RectangleF> Rectangles => rectangles; 
```
тогда да, понятно зачем

PavelLediankin replied:
О, тогда так я и перепишу

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+        private List<RectangleF> rectangles;
+        public List<RectangleF> Rectangles { get => rectangles; }
+        private PointF center;
+        private IEnumerator<PointF> spiralPath;
```
можно сделать readonly

PavelLediankin replied:
+

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+            bitmap.Save(filename);
+        }
+
+        private RectangleF GetRectangle(SizeF rectSize, PointF center)
```
Не понятно зачем нужен этот метод, почему бы сразу не вызывать ```RectangleFExtensions.GetRectangleByCenter(rectSize, center)```
Так ты только излишне плодишь код

PavelLediankin replied:
Согласен, просто не убрал когда выносил метод в Extensions

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+        private RectangleF GetRectangle(SizeF rectSize, PointF center)
+            => RectangleFExtensions.GetRectangleByCenter(rectSize, center);
+
+        private RectangleF GetShiftedRectangle(SizeF rectSize,
```
Очень странный перенос параметров.
Строка получается не такой длинной, и параметры можно вообще не переносить.
Но если уж хочется перенести, то принято разбивать так: каждый новый параметр с новой строки.

PavelLediankin replied:
У меня этот метод немного вылазит за рабочую область, а выделять каждый параметр с новой строки мне не очень нравится. 
Может просто перенести все параметры в новую строку как здесь например ? :
***
ThrowException_WhenPutRectangle_WithNotPositiveSize
            (int width, int height)

dmnovikova replied:
Видимо, у тебя слишком маленькая область, будет лучше, если ты ее поправишь.
Предложенный тобой вариант лучше существующего, но обычно так не пишут.

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+            => RectangleFExtensions.GetRectangleByCenter(rectSize, center);
+
+        private RectangleF GetShiftedRectangle(SizeF rectSize,
+            PointF center, Vector2 offset)
```
С ходу не понятно чем параметр center отличается от поля center?
Из-за этого страдает читабельность кода. Ведь чтобы понять разницу, приходится вчитываться в весь класс, 
Поэтому пересечение в названиях - плохая практика.

PavelLediankin replied:
ок переименую

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+            while (IsRectangleIntersectedByAnother(rect))
+            {
+                spiralPath.MoveNext();
+                var p = spiralPath.Current;
```
Плохое название переменной, оно ничего не несет

PavelLediankin replied:
Я бы не согласился - это Enumerator точек выдающий их в виде спирали, может не лучшее название но вполне понятное

dmnovikova replied:
Давай рассмотрим следующую строчку в отрыве от текущей.
```rect = GetRectangle(rect.Size, p);```
Мы создаем прямоугольник, передаем ему размер и ... что же мы еще передаем?
тут два варианта: искать место определения переменной p (и то, что она строчкой выше нам просто повезло), либо проваливаться внутрь метода _GetRectangle_
и то и другое отнимает время.

А теперь почувствуй разницу, если бы было написано
```rect = GetRectangle(rect.Size, position);```

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+            return ShiftRectangleToCenter(rect);
+        }
+
+        private RectangleF ShiftRectangleToCenter(RectangleF rect)
```
Из-за количества разных действий, получился слишком нагруженный метод.
Я бы советовала добавить пустые строки в некоторых местах метода, чтобы улучшить читаемость.

PavelLediankin replied:
Не уверен что это нужно -  в цикле это не слишком нужно- максимум отделить получение прямоугольника от сдвига точки, а после цикла вообще бессмысленно  - "}\n" на мой взгляд является достаточным разделителем

dmnovikova replied:
Если разделить код на 3 блока (как паттерн AAA в тестах): какие-то подготовительные действия, сам процесс, заявленный в названии функции и возвращаемое значение, то код правда улучшится.

Можешь не верить, но сколько рабочих проектов я не видела, все так делают :)

dmnovikova replied:
Исключение - небольшой метод на 2-3 строки

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+            return shifted;
+        }
+
+        private IEnumerable<PointF> GetSpiralPath()
```
Что будет, если я захочу создать **рядом** новую реализацию спирали (или сделать другую фигуру) и сравнить ее плотность с текущей?

dmnovikova replied:
Подсказка: расширяем ли код?

PavelLediankin replied:
Предлагаете выделить отдельный класс? Я думал об этом но решил что выносить ради пары методов
get  IEnumerable, IEnumerator показалось лишним

dmnovikova replied:
Да, лучше вынести.
Помимо плохой расширяемости, тут еще нарушен Single-responsibility principle, про который я тоже уже упоминала

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+        }
+
+        private bool IsRectangleIntersectedByAnother(RectangleF rect)
+            => rectangles.Where(r => r.IntersectsWith(rect))
```
можно просто ```rectangles.Any(r => r.IntersectsWith(rect))```

PavelLediankin replied:
Не знал, спасибо

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+            {
+                spiralPath.MoveNext();
+                var p = spiralPath.Current;
+                rect = GetRectangle(rect.Size, p);
```
Для чистоты кода лучше, если методы, используемые в текущем будут объявлены после.
Т.е. GetRectangle будет расположен в файле ниже всех своих использований.

PavelLediankin replied:
+

dmnovikova replied:
Будь внимательнее, как минимум еще GetShiftedRectangle попадает под этот же случай

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+        private IEnumerable<PointF> GetSpiralPath()
+        {
+            var a = 0;
+            var k = 0.01;
```
это константа, а не переменная

PavelLediankin replied:
Можно выделить константу, но не уверен что это нужно

dmnovikova replied:
Можно не выделять, а тут же заменить на const 
Согласна, что это мелочь, но подсказка решарпера перед глазами маячит - не приятно

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+
+namespace TagsCloudVisualization
+{
+    public class CircularCloudLayouter
```
Сейчас класс получился перегруженным.
Помимо основной своей задачи - располагать новый прямоугольник, он выполняет дополнительные.
Нарушен Single-responsibility principle (можешь подробнее почитать про SOLID принципы)

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        private List<Size> PutSeveralRectangles
+            (int count, int min = 1, int max = 20)
```
зачем переносить параметры в отдельную строку?

PavelLediankin replied:
Параметры немного уходят за рабочую область - для удобства

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+            return result;
+        }
+
+        private List<Size> PutSeveralRectangles
```
Функция получилась не чистой - помимо того, что она кладет прямоугольники (как заявлено в названии функции), она еще и генерирует их.
Советую погуглить про чистые функции.

PavelLediankin replied:
Вообще я добавил генерацию сюда, как раз чтобы не вызывать каждый раз генерацию списка размеров в тестах, Мне казалось что важнее избавиться от лишнего дублирования

dmnovikova replied:
Чистая функция важнее

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+            // не смог найти способ изменить сообщение теста поэтому пришлось 
+            //добавлять его таким образом
+            TestContext.WriteLine(message);
+            //Assert.IsTrue(false, message);
```
Что делает эта строчка? Выглядит как что-то забытое.
Советую перед отправкой на ревью проводить селф-ревью. И задаваться вопросом, готов ли сейчас твой код к влитию в мастер?

PavelLediankin replied:
На самом деле нет, это другой вариант добавления сообщения к упавшему тесту, но так, как оба варианта мне не нравятся, я решил оставить оба, чтобы получить совет какой лучше (ну или как это можно сделать иначе)

dmnovikova replied:
TestContext.WriteLine(message); норм

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+        public void PlaceRectanglesCloseToCircularForm()
+        {
+            PutSeveralRectangles(400, 18, 20);
+            var center = this.center.ToVector();
```
Опять же, пересечение в названиях

PavelLediankin replied:
+

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+        [Test]
+        public void PlaceRectanglesCloseToCircularForm()
+        {
+            PutSeveralRectangles(400, 18, 20);
```
нужно ли тут указывать максимальное количество?

PavelLediankin replied:
Не понял - прямоугольников, размеров или чего-то еще?

dmnovikova replied:
я про параметр 20

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+        [Test]
+        public void WorkFastEnough()
+        {
+            var layouter2 = new CircularCloudLayouter(new PointF(100, 100));
```
Давай переименуем?

PavelLediankin replied:
WorkFastEnough ?
или layouter2 ?
если 2  - на самом деле я не уверен что он нужен, все таки вряд ли тут требуется настолько большая точность

dmnovikova replied:
layouter2
цифры вообще особо не стоит писать в названиях
лучше указать чем он отличается уже от существующего, или какую-нибудь его особенность
а когда совсем трудно придумать название, пиши other<переменная> или new<переменная> :)

да, можно совсем не создавать его

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+        {
+            PutSeveralRectangles(count);
+            var rectangles = layouter.Rectangles;
+            for (var i = 0; i < rectangles.Count; i++)
```
Это вкусовщина, но лучше оборачивать вложенный for
Почему? Чтобы минимизировать риск ошибки, когда кто-то придет править твой код, и не будет разбираться во всей логике

PavelLediankin replied:
Ну не знаю, в данном случае думаю это не нужно

dmnovikova replied:
Хорошо, как я уже сказала - это дело вкуса

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        [Test]
+        public void PutRectangleToList()
```
Тут и в других тестах не понятно по названию что проверяется

PavelLediankin replied:
+

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+namespace TagsCloudVisualization
+{
+    [TestFixture]
+    public class CircularCloudLayouterShould
```
Когда пишутся тесты на какой-то класс, то они именуются следующим образом: <Название класса>Tests

PavelLediankin replied:
Видел варианты и с Should, но в данном случае возможно и правда стоит поменять название для более простого именования тестов

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+using System.Linq;
+using System.Collections.Generic;
+using FluentAssertions.Extensions;
+using System.Numerics;
```
лишний юзинг

PavelLediankin replied:
+

dmnovikova commented on cs/TagsCloudVisualization/Extensions.cs:
```diff
@@ -0,0 +1,36 @@
+
+
+
+        public static double GetDistanseTo(this Vector2 v, PointF point)
```
опечатка в Distance

dmnovikova replied:
Лучше писать в одном стиле: если ты сокращаешь названия до 1й буквы у одной переменной, делай это и у второй, или наоборот пиши полные у обоих.

PavelLediankin replied:
+ и +

dmnovikova replied:
не поняла

PavelLediankin replied:
плюс в начале строки отмечается как-то странно
  .+ и + (согласен, исправлюсь по обоим комментариям -)

dmnovikova commented on cs/TagsCloudVisualization/Extensions.cs:
```diff
@@ -0,0 +1,36 @@
+            => (point.ToVector() - v).Length();
+    }
+
+    public static class PointFExtensions
```
Давай вынесем это в отдельный класс.
Не бойся делать небольшие файлы. Представь огромный проект, ты найдешь этот код быстрее в файле Extensions или в PointFExtensions?

PavelLediankin replied:
ну хорошо, но все же зная что количество расширений не сильно увеличится я бы не стал это делать (или объединил в одной папке например, чтобы не занимали много места)

dmnovikova replied:
папка - норм
насчет увеличения расширений я бы не зарекалась :)
появляется новый алгоритм выбора позиции - появляются новые расширения

dmnovikova commented on cs/TagsCloudVisualization/Extensions.cs:
```diff
@@ -0,0 +1,36 @@
+﻿using System;
```
лишний юзинг

PavelLediankin replied:
В принципе, конечно можно убрать, но System используется часто

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        /// <summary>
+        /// Я решил сделать визуализацию через тесты, чтобы не выделять отдельный проект
```
Давай подумаем о правильности такого подхода.
Представим, что мы не знакомы с проектом. Знаем лишь только, что он умеет рисовать облако тегов.
И мы захотим нарисовать облако тегов со своими параметрами.
Во-первых, мы не знаем, что именно Layouter рисует облако - в его названии это никак не отображено.
Во-вторых, мы и не подумаем пойти в тесты на этот самый Layouter.

Скорее всего, мы будет искать некий EntryPoint, или Program

PavelLediankin replied:
Я реализую это как библиотеку так что EntryPoint-а нет, и тогда мой подход оправдан - пользователь  проверит (т.к. до того как нарисовать облако ему еще нужно будет его заполнить) публичные методы Layouter-а: увидит PutNextRectangle и Visualise, и поймет где идет отрисовка.
А рисовать именно через тесты я решил для сокращения времени и работы - отдельный проект, который нужен только для того чтобы нарисовать несколько раскладок на мой взгляд лишний

dmnovikova replied:
А нужен ли отдельный проект для отрисовки?
по моему, текущее название - TagsCloudVisualization вполне позволяет делать это в текущем  
не считая того, что ты сделал библиотеку :)

dmnovikova replied:
И раз это библиотека, то в Properties у проекта  должно быть указано, что он является библиотекой
Сейчас там указано, что это консольное приложение

dmnovikova replied:
И еще, опять же вкусовщина, но мне кажется странным тесты пихать в одну dll с библиотекой. Размер бинаря увеличивается, и в целом непонятно зачем их с собой таскать.

PavelLediankin replied:
Вообще странно конечно, я уверен что создавал именно как библиотеку, 
Тогда изменю в Properties,
Сделаю отдельный проект под тесты
А визуализацию оставлю в тестах (потому что мне кажется это должно быть не отдельное приложение, а библиотека)

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+            var path = $"..\\..\\..\\{name}.jpg";
+            var message = $"Test {name} down! +\n" +
+                $"Tag cloud visualization saved to file {path}";
+            layouter.Visualise(path);
```
Что  произойдет, если упадет тест BeEmptyAfterCreation?

PavelLediankin replied:
Пустая картинка видимо, надо будет исправить

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+                return;
+            var name = TestContext.CurrentContext.Test.Name;
+            var path = $"..\\..\\..\\{name}.jpg";
+            var message = $"Test {name} down! +\n" +
```
Первый + попадает в сообщение. Так и должно быть?

PavelLediankin replied:
опечатка

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,104 @@
+            return shifted;
+        }
+
+        private IEnumerable<PointF> GetSpiralPath()
```
Дополнительно: можно добавить тесты на спираль. Поисследовать, на каком расстоянии она создает точки.

PavelLediankin replied:
Если будет время добавлю

dmnovikova commented:
Есть моменты, не привязанные к коду: 

- Лучше создавать отдельную ветку, и пушить изменения в нее.  В рамках этой задачи это не критично, но как правило, мастером пользуются несколько человек. И если каждый будет пушить изменения по своей задаче в мастер, в нем начнется хаос. А задеплоить готовую задачу просто будет невозможно, ведь задача коллеги еще не завершена, и ее изменения никак нельзя на прод! 
- Нужно форматировать файлы (ctrl+alt+L) и пользоваться решарпером.

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouterShould.cs:
```diff
@@ -0,0 +1,170 @@
+
+        private List<Size> GetRandomSizes(int count, int min, int max)
+        {
+            var rnd = new Random();
```
new Random(); дает полный рандом
Одно из требований к хорошему тесту - он должен быть повторяемым.
Если передать в конструктор Random константный seed - проблема порешается.

PavelLediankin replied:
добавлю

PavelLediankin committed:
Изменения в организации

TagsCloudVisualization - тип выходных данных изменен
Выделен проект с тестами

PavelLediankin committed:
Extensions

Extensions разбиты на несколько файлов

PavelLediankin committed:
Выделен класс Spiral

PavelLediankin committed:
TagsCloudVisual Refactoring

PavelLediankin committed:
Refactoring LayouterTests +

 + исправлены пропущенные ошибки в extensions

PavelLediankin commented:
> * Лучше создавать отдельную ветку, и пушить изменения в нее.  В рамках этой задачи это не критично, но как правило, мастером пользуются несколько человек. И если каждый будет пушить изменения по своей задаче в мастер, в нем начнется хаос. А задеплоить готовую задачу просто будет невозможно, ведь задача коллеги еще не завершена, и ее изменения никак нельзя на прод!

Да, я знаю это, но в данном случае я не могу добавить ветку в чужой репозиторий и соответственно смогу пушить только в мастер, а значит и от создания ветки толку нет

PavelLediankin committed:
Забытый момент

улучшено название переменной

dmnovikova commented:
> > * Лучше создавать отдельную ветку, и пушить изменения в нее.  В рамках этой задачи это не критично, но как правило, мастером пользуются несколько человек. И если каждый будет пушить изменения по своей задаче в мастер, в нем начнется хаос. А задеплоить готовую задачу просто будет невозможно, ведь задача коллеги еще не завершена, и ее изменения никак нельзя на прод!
> 
> Да, я знаю это, но в данном случае я не могу добавить ветку в чужой репозиторий и соответственно смогу пушить только в мастер, а значит и от создания ветки толку нет

Форк - это копия. Т.е. по сути это твой личный репозиторий, уже в котором у тебя есть права :)

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+            rectangles.Add(positionedRectangle);
+        }
+
+        public void Visualise(string filename)
```
Давай теперь сделаем отрисовку вариативной.
Например, можно использовать разные цвета для отрисовки и фона, и прямоугольников: они могут быть захардкожены, или указываться пользователем. Можешь придумать свою вариативность, в этом никак не ограничиваю

dmnovikova replied:
А еще если центр облака не совпадает с центром изображения, то информация о центре теряется
Давай также будем отрисовывать центр облака

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+    {
+        private readonly List<RectangleF> rectangles;
+        public IReadOnlyList<RectangleF> Rectangles => rectangles;
+        private readonly PointF layouterCenter;
```
Layouter переводится как "компановщик".
переменная center обозначает центр облака, которые мы формируем, а не центр компановщика
давай вынесем Center и Rectangles в отдельный тип - Cloud

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+            var gr = Graphics.FromImage(bitmap);
+            var pen = new Pen(Color.DarkGreen, 1);
+
+            gr.DrawRectangles(pen, rectangles.ToArray());
```
Сейчас есть проблема: нет масштабируемости.
Т.е., сейчас отрисовка происходит 1:1, если будет слишком много прямоугольников, или их размеры будут слишком большими, то облако не поместится в изображение

dmnovikova replied:
graphic.ScaleTransform должен подойти

dmnovikova commented on cs/TagsCloudVisualisationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,168 @@
+        [Test]
+        public void LayouterShould_PlaceRectangles_CloseToCircularForm()
+        {
+            PutSeveralRectangles(400, 18);
```
Прямоугольники получаются почти квадратными - так слишком просто.
Давай сделаем большой разброс:
от 1 до 20
от 20 до 100
от 100 до 500
чтобы проверить, что алгоритм раскладки действительно создает круглую форму, и это не зависит от размеров прямоугольников

dmnovikova commented on cs/TagsCloudVisualisationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,168 @@
+            layouter.Visualise(filename);
+        }
+
+        private List<Size> GetRandomSizes(int count, int min, int max, int seed = 0)
```
Стоит ли передавать seed параметром, если в итоге ты ни в одном из тестов явно его не указываешь?

dmnovikova replied:
не исправлено

PavelLediankin replied:
Может и не нужно передавать seed , но в случае если будет нужно сравнить несколько различных облаков с одним min, max, но разным seed - тогда для этого все уже готово.

dmnovikova replied:
Ну сейчас это не нужно, а глаз цепляет
Заново переопределить когда это понадобится будет не сложно

dmnovikova commented on cs/TagsCloudVisualisationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,168 @@
+
+            var name = TestContext.CurrentContext.Test.Name;
+            var path = $"..\\..\\..\\{name}.jpg";
+            var message = $"Test {name} down!\n" +
```
Падает тест -> выводится ```Tag cloud visualization saved to file ..\..\..\LayouterShould_PlaceRectangles_CloseToCircularForm.jpg```
Ссылка кликабельна - это хорошо
Но я не вижу где в действительности лежит файл, было бы здорово, если бы выводился абсолютный путь

dmnovikova replied:
Еще давай все изображения складывать в папку images, чтобы не засорять проект

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+        public void Visualise(string filename)
+        {
+            var bitmap = new Bitmap
+                ((int)(layouterCenter.X * 2), (int)(layouterCenter.Y * 2));
```
Что будет, если я укажу центром точку (0, 0) ?

dmnovikova replied:
Так же никто мне не мешает написать так: ```new CircularCloudLayouter(new PointF(-5, 0));```
Давай смотреть на изображение как на систему координат - центр изображения - точка (0, 0)

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+            return shifted;
+        }
+
+        private Vector2 GetNormalToCenter(PointF point)
```
Можно вынести в экстеншн метод

dmnovikova commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace TagsCloudVisualization
+{
+    class Spiral
```
internal
явное лучше, чем неявное

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,92 @@
+
+        private RectangleF ShiftRectangleToCenter(RectangleF rect)
+        {
+            var shifted = new RectangleF();
```
я бы перенесла эту строчку перед while (вместе с инициализацией k)
Иначе ты выделяешь память, а потом в if'у ниже у тебя получается true? и ты выходишь из метода

Предлагаю переписать так:
```
private RectangleF ShiftRectangleToCenter(RectangleF rect)
        {
            var rectCenter = rect.GetCenter();
            var normal = GetNormalToCenter(rectCenter);
            
            if (float.IsNaN(normal.X) || float.IsNaN(normal.Y))
                return rect;
            
            var shifted = new RectangleF();
            var k = 1;
            while (!IsRectangleIntersectedByAnother(rect))
            {
                shifted = rect;
                rect = GetShiftedRectangle(rect.Size, rectCenter, normal * k);
                k += 1;
            }
            
            return shifted;
        }
```

dmnovikova commented on cs/TagsCloudVisualization/Extensions/Vector2Extensions.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System.Drawing;
+using System.Numerics;
+
+namespace TagsCloudVisualization
```
поправь неймспейс
решарпер ведь подказывает...

PavelLediankin committed:
update Visual Test & Refactoring

Изображения перенесены в папку images
Ссылки на изображения созданные в тестах указаны через абсолютный путь
Исправлены ошибки в слове Visualize
Добавлены проверки в тест
LayouterShould_PlaceRectangles_CloseToCircularForm

PavelLediankin committed:
Update Extensions

change namespace for Extension,
add GetNormalToCenter() in Extensions

PavelLediankin committed:
Small Refactoring

PavelLediankin committed:
выделен Cloud, добавлена пользовательская визуализация

PavelLediankin committed:
Visualisation

Custom Visualisation - произвольный центр облака (относительно изображения), произвольные цвета
Рефакторинг

PavelLediankin committed:
Test Refactoring

Чистая функция
закомментирована пользовательская визуализация

PavelLediankin committed:
GetShiftedRectangle смещен ниже

dmnovikova commented on cs/TagsCloudVisualization/Cloud.cs:
```diff
@@ -0,0 +1,100 @@
+                getRectanglesByColorIndex = GetRectanglesByColorIndexDefaultFunc(colors);
+            var gr = Graphics.FromImage(bitmap);
+
+            gr.FillRectangle(new SolidBrush(backgroundColor), 
```
можно проще: graphics.Clear(backgroundColor);

dmnovikova commented on cs/TagsCloudVisualization/Cloud.cs:
```diff
@@ -0,0 +1,100 @@
+            List<Pen> pens,
+            Graphics gr)
+        {
+            if (fillRectangles)
```
Сейчас трудночитаемая логика с построением ```Func<int, List<RectangleF>> getRectanglesByColorIndex```

Можно проще:
```
var color = colors[i % colors.Count];
var br = new SolidBrush(color);
                
if(fillRectangles)
{
    gr.FillRectangle(br, rectangles[i]);
}
else
{
    gr.DrawRectangle(pens[i], rectangles[i]);
}
```

dmnovikova replied:
Это получится, если использовать структуру Rectangle, а не RectangleF
Думаю, мы можем поменять, ведь ты все равно везде указываешь размер как целые числа

PavelLediankin replied:
Кажется RectangleF используется, из-за невозможности получить центр при нечетном размере и трудности размещения прямоугольника в этом случае.
Кроме того предполагается что пользователь сможет задавать эту функцию для другого распределения цветов (условно - один для одного витка спирали, другой для следующего и т.д.)

dmnovikova replied:
Про центр: смещение на 0.5 - не страшно

dmnovikova replied:
Насчет других функций: 

1. сейчас нигде не указано, что пользователь сможет передавать функцию, а значит пока это излишне
2. тогда можно так:
```
var color = myFunc != null 
                        ? myFunc(allRectangles **или другие какие параметры**) 
                        :  colors[i % colors.Count];
var br = new SolidBrush(color);
                
if(fillRectangles)
{
    gr.FillRectangle(br, rectangles[i]);
}
else
{
    gr.DrawRectangle(pens[i], rectangles[i]);
}
```

PavelLediankin replied:
Func<int, List<RectangleF>> getRectanglesByColorIndex - может передаваться как аргумент, максимум можно добавить описание этого аргумента
Насчет центра - может не страшно, но сейчас переделывать  все Size, Rectangle, Point - будет слишком неудобно, К тому же при целых числах, прямоугольники возможно могут пересекаться, так что заменять это я не буду

dmnovikova commented on cs/TagsCloudVisualization/Cloud.cs:
```diff
@@ -0,0 +1,100 @@
+        public void AddRectangle(RectangleF rectangle) 
+            => rectangles.Add(rectangle);
+
+        public void DefaultVisualize(string filename)
```
Что будет, если мы добавим еще один тип облака, например, облако с максимальным количеством тегов (прямоуголников), и его мы тоже захотим отрисовывать?
Помимо того, что у нас будет дублирование кода - нам так же придется заново указывать параметры отрисовки.
Должно ли облако уметь себя рисовать? Или лучше, если будет некий объект, который будет принимать другие объекты для отрисовки?

dmnovikova replied:
У тебя до сих пор облако умеет рисовать само себя.
Это нарушение single responsibility principle!

dmnovikova replied:
К тому же именования DefaultVisualize и CustomVisualize можно свести до одного понятно имени Visualize, с разным набором аргументов

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
```
лишний юзинг, не забывай форматировать файлы!

dmnovikova commented on cs/TagsCloudVisualisationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,202 @@
+            cloud.CustomVisualize(filename, size, colors, background, fillRect);
+        }
+
+        private List<Size> GetRandomSizes(int count, int min, int max, int seed = 0)
```
static

PavelLediankin committed:
мелкие исправления

PavelLediankin committed:
Визуализация в отдельный метод

PavelLediankin committed:
Визуализация в отд метод 2

(Не сохранилось)

dmnovikova commented on cs/TagsCloudVisualization/RectanglesVisualizator.cs:
```diff
@@ -0,0 +1,93 @@
+        }
+
+        public void CustomVisualize(string filename,
+            Size bitmapSize,
```
не обязательно: настройки отрисовки можно объединить в один класс VisualizatorSettings

PavelLediankin replied:
+

dmnovikova commented on cs/TagsCloudVisualization/RectanglesVisualizator.cs:
```diff
@@ -0,0 +1,93 @@
+            List<Pen> pens,
+            Graphics gr)
+        {
+            if (fillRectangles)
```
тут уже точно оборачивать в {}

PavelLediankin replied:
Ну в принципе можно, хотя я бы не сказал что читаемость сильно вырастет

PavelLediankin committed:
Масштабирование

 + небольшие изменения в параметрах визуализации

PavelLediankin committed:
VisSettings + Refactoring

PavelLediankin committed:
убрал юзинги

PavelLediankin committed:
убрана визуализация из облака

PavelLediankin committed:
cloud в параметры визуализации

dmnovikova commented:
Зачтено ✅👍
